opentracing-usecases/src/test/java/io/opentracing/usecases/AutoFinishScopeManagerTest.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2016-2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.usecases;
+
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class AutoFinishScopeManagerTest {
+    private AutoFinishScopeManager autoFinishActivator;
+
+    @Before
+    public void before() throws Exception {
+        autoFinishActivator = new AutoFinishScopeManager();
+    }
+
+    @Test
+    public void missingActiveSpan() throws Exception {
+        Scope missingScope = autoFinishActivator.active();
+        Assert.assertNull(missingScope);
+    }
+
+    @Test
+    public void makeActiveSpan() throws Exception {
+        Span span = Mockito.mock(Span.class);
+
+        try (AutoFinishScopeManager.AutoFinishScope scope = autoFinishActivator.activate(span)) {
+            Assert.assertNotNull(scope);
+            Scope otherScope = autoFinishActivator.active();
+            Assert.assertEquals(otherScope, scope);
+        }
+
+        // Make sure the Span got finish()ed.
+        Mockito.verify(span).finish();
+
+        // And now it's gone:
+        Scope missingScope = autoFinishActivator.active();
+        Assert.assertNull(missingScope);
+    }
+
+    @Test
+    public void deferring() throws Exception {
+        Span span = Mockito.mock(Span.class);
+
+        AutoFinishScopeManager.AutoFinishScope.Continuation continuation = null;
+        {
+            // We can't use 1.7 features like try-with-resources in this repo without meddling with pom details for tests.
+            try (AutoFinishScopeManager.AutoFinishScope scope = autoFinishActivator.activate(span)) {
+                // Take a reference...
+                continuation = scope.defer();
```

Comment:
Not having the `defer` and `Continuation` as part of the API will be a problem, as it means any async based framework instrumentations will be bound to the `AutoFinishScopeManager` implementation.

If we want to reduce the number of components, how about having `Scope.defer()` return a deferred scope which can then be activated in the new context? So the `Scope` is either activated immediately on `startActive()` or explicitly via its `activate()` method - but can only be activated once.